include/sundials/sundials_math.h

```diff
@@ -158,6 +246,128 @@ extern "C" {
+ * SUNRfloor(x) returns the largest integer value not greater than x.
+ * -----------------------------------------------------------------
+ */
+
+#ifndef SUNRfloor
+#if defined(SUNDIALS_DOUBLE_PRECISION)
+#define SUNRfloor(x) (floor((x)))
+#elif defined(SUNDIALS_SINGLE_PRECISION)
+#define SUNRfloor(x) (floorf((x)))
+#elif defined(SUNDIALS_EXTENDED_PRECISION)
+#define SUNRfloor(x) (floorl((x)))
+#else
+#error \
+  "SUNDIALS precision not defined, report to github.com/LLNL/sundials/issues"
+#endif
+#endif
+
+/*
+ * -----------------------------------------------------------------
+ * Function : SUNRround
+ * -----------------------------------------------------------------
+ * Usage : sunrealtype round_x;
+ *         round_x = SUNRround(x);
+ * -----------------------------------------------------------------
+ * SUNRround(x) returns the nearest integer value to x (in floating-point format), 
+ * rounding halfway cases away from zero, regardless of the current rounding mode.
+ * -----------------------------------------------------------------
+ */
+
+#ifndef SUNRround
+#if defined(SUNDIALS_DOUBLE_PRECISION)
+#define SUNRround(x) (round((x)))
+#elif defined(SUNDIALS_SINGLE_PRECISION)
+#define SUNRround(x) (roundf((x)))
+#elif defined(SUNDIALS_EXTENDED_PRECISION)
+#define SUNRround(x) (roundl((x)))
+#else
+#error \
+  "SUNDIALS precision not defined, report to github.com/LLNL/sundials/issues"
+#endif
+#endif
+
+/*
+ * -----------------------------------------------------------------
+ * Function : SUNIceil
+ * -----------------------------------------------------------------
+ * Usage : sunindextype ceil_x;
+ *         ceil_x = SUNIceil(x);
+ * -----------------------------------------------------------------
+ * SUNIceil(x) returns the smallest sunindextype value not less than x.
+ * -----------------------------------------------------------------
+ */
+
+#ifndef SUNIceil
+#if defined(SUNDIALS_DOUBLE_PRECISION)
+#define SUNIceil(x) ((sunindextype)(ceil((x)) + SUN_RCONST(0.5)))
+#elif defined(SUNDIALS_SINGLE_PRECISION)
+#define SUNIceil(x) ((sunindextype)(ceilf((x)) + SUN_RCONST(0.5)))
+#elif defined(SUNDIALS_EXTENDED_PRECISION)
+#define SUNIceil(x) ((sunindextype)(ceill((x)) + SUN_RCONST(0.5)))
+#else
+#error \
+  "SUNDIALS precision not defined, report to github.com/LLNL/sundials/issues"
+#endif
+#endif
+
+/*
+ * -----------------------------------------------------------------
+ * Function : SUNIfloor
+ * -----------------------------------------------------------------
+ * Usage : sunindextype floor_x;
+ *         floor_x = SUNIfloor(x);
+ * -----------------------------------------------------------------
+ * SUNIfloor(x) returns the largest sunindextype value not greater than x.
+ * -----------------------------------------------------------------
+ */
+
+#ifndef SUNIfloor
+#if defined(SUNDIALS_DOUBLE_PRECISION)
+#define SUNIfloor(x) ((sunindextype)(floor((x)) + SUN_RCONST(0.5)))
```

Comment:
What do you think @drreynolds? Where do you think we should document this behavior?

Reply:
I believe that adding 0.5 is the "standard" trick to ensure that floating-point output from `floor` is cast to the correct integer.

SUNDIALS has no documentation for our "math" functions, so I do not have a recommendation on where to put it there.  I recommend that the comments above (line 329) be updated to mention the danger when casting the output of floor to the nearest `sunindextype` for arguments that exceed the representable range of `sunindextype`.

Reply:
The `round` function returns a `double`, while an `int` is expected. Casting a `double` to an `int` will truncate the decimal part. If `round` produces a value slightly less than the expected integer, casting will result in an integer that is one less than desired. To avoid this phenomena, we add 0.5 after rounding and before casting.

Reply:
I may be off on this, but if the algorithm to compute the number of required stages for a given step size `h` and spectral radius `rho` is updated to do the comparisons/calculations in floating-point arithmetic, and to then only cast the final result to `int`, do we even need these new integer-valued functions (e.g., `SUNIfloor`)?

Reply:
No matter at which stage we do, we have to cast `ss` to an integer. Can you check the current commit? I think your suggestion will help with limiting `ss` so that we will not encounter too big integer issues.

Reply:
I like the change.  My only recommendation is to use `step_mem->req_stages = SUNIceil(ss);` so that we are certain to use a "stable" number of stages.

Reply:
I ceiled the number once on the top to find `ss`, so ceiling one more time might end up increasing it by one for the same `double` argument type reasoning. `round` seemed to be more appropriate option for me.